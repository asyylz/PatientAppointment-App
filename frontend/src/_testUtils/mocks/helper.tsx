import { within } from '@testing-library/react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { Store } from '@reduxjs/toolkit';
import React from 'react';

export const toContainRole = async (
  container: HTMLElement,
  role: string,
  quantity = 1
): Promise<{ pass: boolean; message: () => string }> => {
  const elements = await within(container).findAllByRole(role);


  if (elements.length === quantity) {
    return {
      pass: true,
      message: () => '', // return an empty string when the test passes
    };
  }

  return {
    pass: false,
    message: () =>
      `Expected to find ${quantity} ${role} elements. Found ${elements.length} instead.`,
  };
};

export const renderComponent = (
  component: React.ReactElement,
  store: Store,
  options?: RenderOptions
) => {
  return render(
    <Provider store={store}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>,
    options
  );
};

// export const toContainRole = (
//   container: HTMLElement,
//   role: string,
//   quantity = 1
// ): { pass: boolean; message: () => string } => {
//   const elements = within(container).getAllByRole(role);
// console.log(elements.length)
//   if (elements.length === quantity) {
//     return {
//       pass: true,
//       message: () => '', // return an empty string when the test passes
//     };
//   }

//   return {
//     pass: false,
//     message: () =>
//       `Expected to find ${quantity} ${role} elements. Found ${elements.length} instead.`,
//   };
// };
