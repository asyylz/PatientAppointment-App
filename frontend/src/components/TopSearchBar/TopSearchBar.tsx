import classes from './TopSearchBar.module.css';
import { useSelector, useDispatch } from 'react-redux';
import { AppDispatch } from '../../store/index';
import { useEffect, useState } from 'react';
import { setSearch } from '../../store/search-slice/search-slice';
import { performLogout } from '../../store/currentUser-slice/currentUser-slice';

const TopSearchBar: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();
  const [searchInput, setSearchInput] = useState('');

  const { token, userData } = useSelector(
    (state: RootState) => state.currentUser
  );

  useEffect(() => {
    const timeout = setTimeout(() => dispatch(setSearch(searchInput)), 500);
    return () => {
      clearTimeout(timeout);
    };
  }, [dispatch, searchInput]);

  const handleLogout = async () => {
    await dispatch(performLogout());
  };

  return (
    <>
      <div className={classes['top__bar']}>
        <div className={classes['top__bar--search']}>
          <input
            className={classes['top__bar--search-input']}
            type="text"
            name="search"
            placeholder="Search here"
            onChange={(e) => setSearchInput(e.target.value)}
          />

          <div className={classes['top__bar--search-icon']}>
            <i className="fas fa-search"></i>
          </div>
        </div>

        {token && userData && (
          <div className={classes['top__bar--search-user']}>
            <h5>
              {userData?.role === 'doctor'
                ? `Dr. ${userData?.name}`
                : userData?.name}
            </h5>
            <i className="fas fa-bell"></i>
            <img src={userData.image} alt="User" />
            <button onClick={handleLogout}>Logout</button>
          </div>
          // ) : (
          //   <button onClick={() => navigate('/auth')}>Login</button>
        )}
      </div>
    </>
  );
};

export default TopSearchBar;
