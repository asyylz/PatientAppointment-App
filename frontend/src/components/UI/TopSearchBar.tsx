import { Link } from 'react-router-dom';
import classes from './TopSearchBar.module.css';
import { useSelector } from 'react-redux';

const TopSearchBar: React.FC = () => {
  const {
    entities: { token, data },
    status,
    error,
  } = useSelector((state: RootState) => state.currentUser);

  return (
    <>
      <div className={classes.topBar}>
        <div className={classes.search}>
          <input type="text" name="search" placeholder="search here" />
          <label htmlFor="search">
            <div className={classes.searchIcon}>
              <i className="fas fa-search"></i>
            </div>
          </label>
        </div>

        {token && data ? (
          <Link to="/logout" className={classes.user}>
            <i className="fas fa-bell"></i>
            <img src="./public/doctor.png" alt="" />
          </Link>
        ) : (
          <Link className={classes.login} to="/auth">
            Login
          </Link>
        )}
      </div>
    </>
  );
};

export default TopSearchBar;
