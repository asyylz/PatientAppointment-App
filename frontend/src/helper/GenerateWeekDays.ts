// Function to get day name from index (0 to 6)
const getDayName = (index: number) => {
  const days = [
    'Sunday',
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
  ];
  return days[index];
};

// Function to generate dates dynamically
export const generateDates = (numberOfDays: number) => {
  const dates = [];
  const today = new Date(); // Get today's date

  for (let i = 0; i < numberOfDays; i++) {
    const date = new Date(today.getTime() + i * 24 * 60 * 60 * 1000); // Add i days to today
    const dayName = getDayName(date.getDay()); // Get day name (e.g., "Monday")

    // Format date as "dd/mm/yyyy"
    const formattedDate = `${date.getDate()}/${
      date.getMonth() + 1
    }/${date.getFullYear()}`;

    // Push formatted date and day name to array
    dates.push({
      date: formattedDate,
      day: dayName,
    });
  }

  return dates;
};

// Example usage: Generate dates for the next 7 days
// const numberOfDays = 7;
// const dates = generateDates(numberOfDays);

// console.log(dates);
