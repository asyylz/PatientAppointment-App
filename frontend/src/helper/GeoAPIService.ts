// GeoAPIService.ts
import axios from 'axios';

const GEO_API_URL =
  'https://api.openweathermap.org/data/2.5/?appid=${apiKey}';
const API_KEY = 'b8b2bc6cc3def4c8452b6812772a682f'; // Replace with your GeoAPI key

interface GeoAPIResponse {
  features: {
    properties: {
      formatted: string;
      street: string;
      city: string;
      state: string;
      postcode: string;
      country: string;
    };
  }[];
}

export const getAddressFromCoordinates = async (lat: number, lon: number) => {
  try {
    const response = await axios.get<GeoAPIResponse>(`${GEO_API_URL}`, {
      params: {
        lat,
        lon,
        apiKey: API_KEY,
      },
    });
    console.log(response.data.features[0].properties);
    return response.data.features[0].properties;
  } catch (error) {
    console.error('Error fetching address:', error);
    throw new Error('Failed to fetch address');
  }
};
