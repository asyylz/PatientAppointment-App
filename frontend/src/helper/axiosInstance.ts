import axios, { AxiosInstance } from 'axios';
import store from './../store/index'; // Adjust the path to your store file

export const createAxiosInstance = (): AxiosInstance => {
  const { token } = store.getState().currentUser;
  console.log(token);
  console.log('asiye')

  return axios.create({
    baseURL: import.meta.env.BASE_URL,
    headers: { Authorization: `Bearer ${token}` },
  });
};
