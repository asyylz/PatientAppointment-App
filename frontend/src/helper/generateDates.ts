const days = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
];

export const getWeekDatesFromToday = () => {
  const today = new Date();
  const currentDayIndex = today.getDay(); // 0 is Sunday, 1 is Monday, etc.
  const startOfWeek = new Date(today);

  // Adjust to start the week from the current day
  startOfWeek.setDate(
    today.getDate() - (currentDayIndex === 0 ? 6 : currentDayIndex - 1)
  );

  // Function to add days to a date
  const addDays = (date: Date, days: number) => {
    const newDate = new Date(date);
    newDate.setDate(date.getDate() + days);
    return newDate;
  };

  const weekDates = days.map((day, index) => {
    const date = addDays(startOfWeek, index);
    const formattedDate = `${date.getDate().toString().padStart(2, '0')}/${(
      date.getMonth() + 1
    )
      .toString()
      .padStart(2, '0')}/${date.getFullYear()}`;
    return { date: formattedDate, day };
  });

  return weekDates;
};

export const convertDateAndTimeStringToDate = (
  dateString: string,
  timeString: string
): Date => {
  // Split time string into hours and minutes
  const [hours, minutes] = timeString.split(':').map(Number);

  // Define a variable to hold the new Date
  let newDate: Date;

  // Check the format of the date string and create a new Date object accordingly
  if (dateString.includes('/')) {
    const [day, month, year] = dateString.split('/').map(Number);
    newDate = new Date(year, month - 1, day, hours + 1, minutes, 0);
  } else {
    const [year, month, day] = dateString.split('-').map(Number);
    newDate = new Date(year, month - 1, day, hours + 1, minutes, 0);
  }

  return newDate;
};

export const formatDateForInput = (dateString: string): string => {
  //console.log(dateString);
  const parts = dateString.split('/');
  return `${parts[2]}-${parts[1]}-${parts[0]}`;
};

export const getDateFromDateString = (dateString: string): string => {
  console.log(dateString);
  const part = dateString.split('T');
  const parts = part[0].split('-');
  return `${parts[2]}/${parts[1]}/${parts[0]}`;
};

export const convertToDateandDateString = (
  availabilityDay: string,
  availabilityTime: string
) => {
  // Map days of the week to their index
  const daysOfWeek = {
    Sunday: 0,
    Monday: 1,
    Tuesday: 2,
    Wednesday: 3,
    Thursday: 4,
    Friday: 5,
    Saturday: 6,
  };

  // Get the current date
  const currentDate = new Date();
  // Get the current week's Monday
  const currentMonday = new Date(
    currentDate.setDate(currentDate.getDate() - currentDate.getDay() + 1)
  );

  // Adjust the day of the week to get the specific day's date
  const availabilityDate = new Date(currentMonday);
  availabilityDate.setDate(
    currentMonday.getDate() +
      (Object.keys(daysOfWeek).indexOf(availabilityDay) - 1 || 0)
  );

  // Combine the availability date and time into a single string
  const availabilityDateTimeString = `${
    availabilityDate.toISOString().split('T')[0]
  }T${availabilityTime}:00.000Z`;

  // Create a Date object for the availability
  const availabilityDateTime = new Date(availabilityDateTimeString);

  //console.log(availabilityDateTimeString);
  // Create a Date object for the current date and time
  const currentDateTime = new Date();

  // Compare the availability date and time with the current date and time
  return { availabilityDateTime, availabilityDateTimeString };
};

/* ---------------------- separate ---------------------- */

export const formatDateForUI = (isoString: string) => {
  console.log(isoString);
  // Extract the date part from the ISO string
  const datePart = isoString.split('T')[0];
  const [year, month, day] = datePart.split('-').map(Number);

  // Create a date object using the extracted date components
  const date = new Date(Date.UTC(year, month - 1, day)); // Date.UTC to avoid time zone conversion

  const options = { day: '2-digit', month: '2-digit', year: 'numeric' };
  const formattedDate = date.toLocaleDateString('en-GB', options); // Format as DD/MM/YYYY

  const daysOfWeek = [
    'Sunday',
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
  ];
  const dayOfWeek = daysOfWeek[date.getUTCDay()]; // Get the day of the week

  return `${formattedDate} ${dayOfWeek}`;
};
