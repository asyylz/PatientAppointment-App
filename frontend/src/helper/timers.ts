import {
  checkTokenExpiration,
  TOKEN_CHECK_INTERVAL,
} from '../hooks/axiosInterceptors';

// Global variable to hold the interval ID
let intervalId: NodeJS.Timeout | null = null;

const startTokenCheckInterval = () => {
  // Clear existing interval if it exists
  if (intervalId) {
    clearInterval(intervalId);
    console.log('Cleared existing interval with ID:', intervalId);
  }

  // Start a new interval
  intervalId = setInterval(checkTokenExpiration, TOKEN_CHECK_INTERVAL);
  console.log('Set new interval with ID:', intervalId);
};

const stopTokenCheckInterval = () => {
  if (intervalId) {
    clearInterval(intervalId);
    console.log('Stopped interval with ID:', intervalId);
    intervalId = null; // Ensure to nullify the interval ID
  }
};

export { startTokenCheckInterval, stopTokenCheckInterval };
