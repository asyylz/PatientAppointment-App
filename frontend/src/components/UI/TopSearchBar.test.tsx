import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { combineReducers } from 'redux';
import { configureStore } from '@reduxjs/toolkit/react';
import TopSearchBar from './TopSearchBar';
import searchReducer from '../../store/search-slice';
import currentUserReducer from '../../store/currentUser-slice';
import { act } from 'react';
import * as router from 'react-router-dom';
import { toastSuccessNotify } from '../../helper/ToastNotify';

jest.mock('../../helper/ToastNotify', () => ({
  toastSuccessNotify: jest.fn(),
}));
jest.mock('redux-persist', () => {
  const actual = jest.requireActual('redux-persist');
  return {
    ...actual,
    persistReducer: jest.fn().mockImplementation((_config, reducer) => reducer),
    // persistStore: jest.fn().mockReturnValue({
    //   // Add any other methods you need to mock
    // }),
  };
});
// Mock useNavigate
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));
// Start store with initial state for `currentUser`
const initialCurrentUserState: CurrentUser = {
  status: 'login success',
  token:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY2NzM2NjJmYmQ0MmE5NjZiNzVkZWM5MiIsImlhdCI6MTcyMzc1NDAzOSwiZXhwIjoxNzIzNzU1ODM5fQ.4Nmyge8591WiGN3wEijyn-9Tx7A0GCBu5w2a0CgVaQs',
  userData: {
    _id: '6673662fbd42a966b75dec92',
    name: 'Asiye',
    email: 'alice@test.com',
    role: 'patient',
    __v: 0,
    DOB: '1986-01-22T00:00:00.000Z',
    image:
      'https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1722605356602-asiye%20serife.jpeg',
    policy: true,
  },
  error: null,
};

const mockAllReducers = combineReducers({
  currentUser: currentUserReducer,
  search: searchReducer,
  // other reducers...
});

const store = configureStore({
  reducer: mockAllReducers,
  preloadedState: {
    currentUser: initialCurrentUserState,
  },
});

describe('TopSearchBar Component', () => {
  const navigate = jest.fn();
  beforeEach(() => {
    jest.spyOn(store, 'dispatch');
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
    render(
      <Provider store={store}>
        <MemoryRouter>
          <TopSearchBar />
        </MemoryRouter>
      </Provider>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <TopSearchBar />
        </MemoryRouter>
      </Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
    expect(screen.getByText('Asiye'))
  });
  /* -------------------------- - ------------------------- */
  it('Search input typed and trigger departments to be filtered', async () => {
    const searchInput = screen.getByPlaceholderText('search here');
    fireEvent.change(searchInput, { target: { value: 'Anesthesiology' } });

    expect(store.dispatch).not.toHaveBeenCalledWith({
      type: 'search/setSearch',
      payload: 'Anesthesiology',
    }); // time for debounce, so  no call yet

    await waitFor(() => {
      expect(store.dispatch).toHaveBeenCalledWith({
        type: 'search/setSearch',
        payload: 'Anesthesiology',
      });
    });

    // Here we expilicitly update state
    await act(async () => {
      store.dispatch({
        type: 'search/setSearch',
        payload: 'Anesthesiology',
      });
    });
    expect(store.getState().search).toEqual('Anesthesiology');
  });
  /* -------------------------- - ------------------------- */
  it('--Logout button should navigate user to home page', async () => {
    const logoutButton = screen.getByText('Logout');
    fireEvent.click(logoutButton);

    // Here we expilicitly update logout
    await act(async () => {
      store.dispatch({
        type: 'currentUser/logout/fulfilled',
      });
    });
    await waitFor(() => {
      expect(store.dispatch).toHaveBeenCalledWith({
        type: 'currentUser/logout/fulfilled',
      });
    });

    const message = 'Successfully logout!';
    toastSuccessNotify(message);
    expect(toastSuccessNotify).toHaveBeenCalledWith(message);
    expect(navigate).toHaveBeenCalledWith('/');
  });
});
