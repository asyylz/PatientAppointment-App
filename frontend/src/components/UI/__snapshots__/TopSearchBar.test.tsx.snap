// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopSearchBar Component 1--Matches the DOM snapshot 1`] = `
<div
  className="top__bar"
>
  <div
    className="top__bar--search"
  >
    <input
      className="top__bar--search-input"
      name="search"
      onChange={[Function]}
      placeholder="search here"
      type="text"
    />
    <div
      className="top__bar--search-icon"
    >
      <i
        className="fas fa-search"
      />
    </div>
  </div>
  <div
    className="top__bar--search-user"
  >
    <h5>
      Asiye
    </h5>
    <i
      className="fas fa-bell"
    />
    <img
      alt="User"
      src="https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1722605356602-asiye%20serife.jpeg"
    />
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</div>
`;
